### PR TITLE
bluetooth: controller: Auto-setup master/slave role count

### PR DESCRIPTION
**Pull request by thomasstenersen**
Remove explicit setting for BLECTRL_MASTER_COUNT, it's now set imlicitly
as BT_MAX_CONN - BLECTRL_SLAVE_COUNT.

The BLECTRL_SLAVE_COUNT is set to:
- BT_CONN_MAX if !BT_CENTRAL
- 1 if BT_CENTRAL && BT_PERIPHERAL
- 0 otherwise

Furthermore, the BLECTRL_SLAVE_COUNT is range limited to BT_MAX_CONN,
meaning Kconfig will reset to a default value if a config outside the
range is set.

@joerchan @rakons @grochu : A better fix than #1406 ?

### CODE REVIEW

File: subsys/bluetooth/controller/hci_driver.c [hunk 1]
```diff
@@ -32,6 +32,20 @@ static K_THREAD_STACK_DEFINE(recv_thread_stack, CONFIG_BLECTLR_RX_STACK_SIZE);
...
+
+/* It should not be possible to set CONFIG_BLECTRL_SLAVE_COUNT larger than
+ * CONFIG_BT_MAX_CONN. Kconfig should make sure of that, this assert is to
+ * verify that assumption. */
+BUILD_ASSERT(CONFIG_BLECTRL_SLAVE_COUNT <= CONFIG_BT_MAX_CONN);
+
+#define BLECTRL_MASTER_COUNT (CONFIG_BT_MAX_CONN - CONFIG_BLECTRL_SLAVE_COUNT)
+
+BUILD_ASSERT((IS_ENABLED(CONFIG_BT_CENTRAL) && (CONFIG_BLECTRL_SLAVE_COUNT > 0)) ||
+			 !IS_ENABLED(CONFIG_BT_CENTRAL));
```
Comment (rakons) [inline subsys/bluetooth/controller/hci_driver.c]: You really do not need !IS_ENABLED part here - it does not change anything.

File: subsys/bluetooth/controller/hci_driver.c [hunk 2]
```diff
@@ -32,6 +32,20 @@ static K_THREAD_STACK_DEFINE(recv_thread_stack, CONFIG_BLECTLR_RX_STACK_SIZE);
...
+
+BUILD_ASSERT((IS_ENABLED(CONFIG_BT_PERIPHERAL) && (BLECTRL_MASTER_COUNT > 0)) ||
+			 !IS_ENABLED(CONFIG_BT_PERIPHERAL));
```
Comment (rakons) [inline subsys/bluetooth/controller/hci_driver.c]: !IS_ENABLED part does not change the result.

**Comment by rakons** [comment]
> A better fix than #1406 ?

I have no strong opinion about witch one is better. Lets wait for the rest colleagues to visit.

File: subsys/bluetooth/controller/hci_driver.c (hunk 1 above)
Comment (thomasstenersen) [reply subsys/bluetooth/controller/hci_driver.c]: It does. If you remove it and have `CONFIG_BT_CENTRAL=n` and `CONFIG_BLECTRL_SLAVE_COUNT=0`, you will get

```c
BUILD_ASSERT( false && (0 > 0) );
```
resulting in `BUILD_ASSERT(false)` and a build assert. 

With the `!IS_ENABLED`:

```c
BUILD_ASSERT( false && (0 > 0) || !false );
```
resulting in `BUILD_ASSERT(true)` and no assert.

File: subsys/bluetooth/controller/Kconfig [hunk 3]
```diff
@@ -6,19 +6,24 @@
...
+	default 1 if BT_PERIPHERAL
+	default 0
+	range 0 BT_MAX_CONN
 	help
 	  Number of concurrent slave roles defines how many simultaneous
 	  connections can be created with the device working as a slave.
+	  NOTE: the number of master roles is defined as
+	  BT_MAX_CONN - BLECTRL_SLAVE_COUNT
+
+config BLECTRL_ERROR_ON_CONN_MISMATCH
```
Comment (grochu) [inline subsys/bluetooth/controller/Kconfig]: Where is this check implemented in the code?

File: subsys/bluetooth/controller/hci_driver.c (hunk 2 above)
Comment (thomasstenersen) [reply subsys/bluetooth/controller/hci_driver.c]: It does, see reasoning in other comment.

**Comment by grochu** [comment]
> A better fix than #1406 ?
Yes, it goes in good direction, as in this PR it won't be possible to auto-generate an inconsistent conn config.

File: subsys/bluetooth/controller/hci_driver.c (hunk 1 above)
Comment (thomasstenersen) [reply subsys/bluetooth/controller/hci_driver.c]: @grochu, you're right :man_facepalming:

File: subsys/bluetooth/controller/Kconfig (hunk 3 above)
Comment (thomasstenersen) [reply subsys/bluetooth/controller/Kconfig]: This should have been removed, I had another check that was removed yesterday :+1:

File: subsys/bluetooth/controller/hci_driver.c (hunk 1 above)
Comment (rakons) [reply subsys/bluetooth/controller/hci_driver.c]: Right, but it still can be simplified to:
```
BUILD_ASSERT(!IS_ENABLED(CONFIG_BT_CENTRAL) || (CONFIG_BLECTRL_MASTER_COUNT > 0))
```

**Comment by thomasstenersen** [comment]
Pushed fixes for checkpatch and @grochu's comments. To avoid noise and confusion, I'll will rebase/squash when changes are approved.

File: subsys/bluetooth/controller/hci_driver.c (hunk 2 above)
Comment (rakons) [reply subsys/bluetooth/controller/hci_driver.c]: Yes. But still can be simplified like in other comment.

File: subsys/bluetooth/controller/hci_driver.c (hunk 1 above)
Comment (thomasstenersen) [reply subsys/bluetooth/controller/hci_driver.c]: Done.

**Comment by thomasstenersen** [comment]
@pdunaj : FYI, the explicit setting of master/slave count has been removed from nrf_desktop samples as they are either central or peripheral.

**Pull request merged**